### PR TITLE
Creating a Fully Managed Beam Streaming System with Flink Runner on Kubernetes - Part 3

### PR DESCRIPTION
**Pull request by talatuyarer**
Hi,

This is last part of our blog series. Please review it but not merge it. I would like to open source our autoscaler code too. 

Thanks

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by talatuyarer** [comment]
@damccorm Do you have any idea how i can use markdown's expression feature ?

**Comment by damccorm** [comment]
> @damccorm Do you have any idea how i can use markdown's expression feature ?

Not 100% sure which feature you're asking about, but its possible its not supported if just using markdown isn't getting the job done. The website is built on top of Hugo with the Docsy theme - https://github.com/apache/beam/blob/master/website/README.md - not sure how complete that is

**Comment by damccorm** [comment]
Staged website for reviews - https://apache-beam-website-pull-requests.storage.googleapis.com/29860/blog/apache-beam-flink-and-kubernetes-part3/index.html

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
+---
+title:  "Behind the Scenes: Crafting an Autoscaler for Apache Beam in a High-Volume Streaming Environment"
+date:   2023-12-21 09:00:00 -0400
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:3]: Leaving a comment to remind us to update this before merging

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+### Introduction to the Design of Our Autoscaler for Apache Beam Jobs
+
+Welcome to the third and final part of our blog series on building a scalable, self-managed streaming infrastructure with Beam and Flink. In our previous post, we delved into the scale of our streaming platforms, highlighting our capacity to manage over 40,000 streaming jobs and process upwards of 10 million events per second. This impressive scale sets the stage for the challenge we address today: the intricate task of resource allocation in a dynamic streaming environment.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:24]: "In our previous post" - would benefit for an inline link to the post

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+In this blog post [Talat Uyarer (Architect / Senior Principal Engineer)](https://www.linkedin.com/in/talatuyarer/), [Rishabh Kedia (Principal Engineer)](https://www.linkedin.com/in/rishabhkedia/) describe more details about our Autoscaler. Imagine a scenario where your streaming system is inundated with fluctuating workloads. Our case presents a unique challenge, as our customers, equipped with firewalls distributed globally, generate logs at various times of the day. This results in workloads that not only vary by time but also escalate over time due to changes in settings or the addition of new cybersecurity solutions from PANW. Furthermore, updates to our codebase necessitate rolling out changes across all streaming jobs, leading to a temporary surge in demand as the system processes unprocessed data.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:26]:
```suggestion
In this blog post [Talat Uyarer (Architect / Senior Principal Engineer)](https://www.linkedin.com/in/talatuyarer/) and [Rishabh Kedia (Principal Engineer)](https://www.linkedin.com/in/rishabhkedia/) describe more details about our Autoscaler. Imagine a scenario where your streaming system is inundated with fluctuating workloads. Our case presents a unique challenge, as our customers, equipped with firewalls distributed globally, generate logs at various times of the day. This results in workloads that not only vary by time but also escalate over time due to changes in settings or the addition of new cybersecurity solutions from PANW. Furthermore, updates to our codebase necessitate rolling out changes across all streaming jobs, leading to a temporary surge in demand as the system processes unprocessed data.
```

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes-part3/resource-allocation.png"
+alt="Resource Allocation">
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:30]: Nit: I think a more accurate/useful graphic might have some overlap between the orange and blue lines for the underprovisioned half. It is pretty rare to have a pipeline or service that is underprovisioned to the point of _never_ having resources, compared to not having enough resources to handle spikes

Comment (damccorm) [reply website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:30]: The graphic as drawn will never catch up

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+Remember, we're aiming for a universal solution applicable to any workload and pipeline. While specific problems might benefit from unique signals, our focus here is on creating a one-size-fits-all approach.
+
+In Flink, tasks form the basic execution unit and consist of one or more operators, such as map, filter, or reduce. Flink optimizes performance by chaining these operators into single tasks when possible, minimizing overheads like thread context switching and network I/O. Your pipeline, when optimized, turns into a directed acyclic graph of stages, each processing elements based on your code. Don't confuse stages with physical machines — they're separate concepts. In Our job we mesuare backlog information by using Apache Beam's [`backlog_bytes` and `backlog_elements`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java#L32) metrics.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:53]:
```suggestion
In Flink, tasks form the basic execution unit and consist of one or more operators, such as map, filter, or reduce. Flink optimizes performance by chaining these operators into single tasks when possible, minimizing overheads like thread context switching and network I/O. Your pipeline, when optimized, turns into a directed acyclic graph of stages, each processing elements based on your code. Don't confuse stages with physical machines — they're separate concepts. In our job we measure backlog information by using Apache Beam's [`backlog_bytes` and `backlog_elements`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/SourceMetrics.java#L32) metrics.
```

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+Downscaling is all about reducing resources without affecting the quality of service. It's about ensuring that we're not overstaffed when the rush hour is over.
+
+#### **Summary: A Recipe for Effective Scaling**
+
+In summary, our scaling policy is for scale up, we first ensure that the time to drain the backlog is beyond the threshold (120s) or the cpu is above the threshold (90%)
+
+Increasing Backlog aka Backlog Growth > 0 :
+
+$`Worker_require = Worker_current \frac{Backlog Growth + Throughput}{Throughput}`$
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:124]: Ah, assuming this is what you were talking about - I'm guessing this just isn't supported; you could take a look through other blogs or website areas and see how they handle it. I'm guessing they just triple backtick

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+To scale down, we need to ensure the machine utilization is low (< 70%) and there is no backlog growth and current time to drain backlog is less than the limit (10s)
+
+So the only driving factor to calculate the required resources after a scale down is CPU
+
+$`CPURate_desired = \frac{Worker_current}{Worker_new} CpuRate_current`$
+
+## Executing Autoscaling Decision
+
+In Our Setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Adviced most of Flink documentation We set per vCPU one slot. Most of Our jobs uses 1 vCPU 4GB Memory combination for TaskManager.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:142]:
```suggestion
In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.
```

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+Parallelism can only be influenced by setting a maximum for pipelines, which the scheduler will honor. Our maxParallelism is limited by the total count of partitions that the pipeline will process, as well as by the job itself. We cap the maximum number of TaskManagers with maxWorker count and control the job's key count in shuffle by setting maxParallelism. Additionally, we set maxParallelism per pipeline to manage pipeline parallelism. The job cannot exceed the job's maxParallelism in terms of workers.
+
+After autoscaler analysis, we will tag if the job needs to be scaled up, no action or scaled down. To interact with the job, we use a library we have built over Flink Kubernetes Operator. This library allows us to interact with our flink jobs via a simple java method call. Library will convert our method call to a kubernetes command.
+
+In the kubernetes world, the call will look like this for a scale up:
+
+`kubectl scale flinkdeployment job-name --replicas=100`
+
+Rest of thing will be handled by Apache Flink.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:162]:
```suggestion
Apache Flink will handle the rest of the work needed to scale up.
```

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md (hunk 4 above)
Comment (talatuyarer) [reply website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:30]: I updated the image

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 9]
```diff
@@ -121,25 +124,33 @@ In summary, our scaling policy is for scale up, we first ensure that the time to
...
 
-$`CPURate_desired = \frac{Worker_current}{Worker_new} CpuRate_current`$
+```math
+CPURate_{desired} = \frac{Worker_{current}}{Worker_{new}} CpuRate_{current}
+```
 
 ## Executing Autoscaling Decision
 
-In Our Setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Adviced most of Flink documentation We set per vCPU one slot. Most of Our jobs uses 1 vCPU 4GB Memory combination for TaskManager.
+In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As Advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:153]:
```suggestion
In our setup we use Reactive Mode which uses Adaptive Scheduler and Declarative Resources manager. We wanted to align resources with slots. As advised in most of the Flink documentation we set one per vCPU slot. Most of our jobs use 1 vCPU 4GB Memory combination for TaskManager.
```

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+src="/images/blog/apache-beam-flink-and-kubernetes-part3/flink-operator-chaining.png"
+alt="Apache Beam Pipeline Optimization by Apache Flink">
+
+##### **Upscaling Signals**
+
+##### *Backlog Growth*
+Let’s take a practical example. Consider a pipeline reading from Kafka, where different operators handle data parsing, formatting, and accumulation. The key metric here is throughput — how much data each operstor processes over time. But throughput alone isn't enough. We need to examine the queue size or backlog at each operator. A growing backlog indicates we're falling behind. We measure this as backlog growth — the first derivative of backlog size over time, highlighting our processing deficit.
+
+```math
+Backlog Growth = \frac{d(Backlog)}{dt}
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:65]: Seems like this is close to what we want, but it is still not quite right - https://apache-beam-website-pull-requests.storage.googleapis.com/29860/blog/apache-beam-flink-and-kubernetes-part3/index.html

Comment (talatuyarer) [reply website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:65]: Yes. According docsy's document it should work but it does not https://www.docsy.dev/docs/adding-content/diagrams-and-formulae/

Yesterday I tried to build on my local. But I am getting this error. I could not investigate more.

```> Task :website:initGitSubmodules FAILED
fatal: detected dubious ownership in repository at '/opt'
To add an exception for this directory, call:

        git config --global --add safe.directory /opt
```
        
 Do you have any idea how can I build the website ?

Comment (damccorm) [reply website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:66]: Have you tried to run the suggested command? `git config --global --add safe.directory /opt`

Unless you're operating on a shared drive that should be safe - https://stackoverflow.com/questions/72978485/git-submodule-update-failed-with-fatal-detected-dubious-ownership-in-repositor - it is a git issue, not an issue with the website building itself

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 11]
```diff
@@ -1,6 +1,7 @@
 ---
 title:  "Behind the Scenes: Crafting an Autoscaler for Apache Beam in a High-Volume Streaming Environment"
 date:   2023-12-21 09:00:00 -0400
+math: true
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:4]: Looks like this did not work. Could you try writing the equations in ways that don't require this special math block?

**Comment by talatuyarer** [comment]
@damccorm Looks like formulas are too big. Tonight I will fix them Tomorrow we can publish it if you are available too

**Comment by talatuyarer** [comment]
@damccorm It looks to me too. If you think it is ok lets publish it.

File: website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md [hunk 12]
```diff
@@ -173,6 +173,18 @@ In the kubernetes world, the call will look like this for a scale up:
...
+
+1. **Max Parallelism** sets an upper limit on how much a job can scale out, ensuring that state can be redistributed across a larger or smaller number of nodes without exceeding predefined boundaries. This is crucial for autoscaling because it allows Flink to manage state effectively, even as the number of task slots changes to accommodate varying workloads.
+2. **Checkpointing** is at the heart of Flink's fault tolerance mechanism, periodically saving the state of each job to a durable storage(In our case it is GCS bucket). In an autoscaling scenario, checkpointing enables Flink to recover to a consistent state after scaling operations. When the system scales out (adds more resources) or scales in (removes resources), Flink can restore the state from these checkpoints, ensuring data integrity and processing continuity without losing critical information. In scale down or up situations there could be a moment to reprocess data from last checkpoint. To reduce that amount we reduce checkpointing interval to 10 seconds.
```
Comment (damccorm) [inline website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part3.md:181]:
```suggestion
2. **Checkpointing** is at the heart of Flink's fault tolerance mechanism, periodically saving the state of each job to a durable storage (in our case it is GCS bucket). In an autoscaling scenario, checkpointing enables Flink to recover to a consistent state after scaling operations. When the system scales out (adds more resources) or scales in (removes resources), Flink can restore the state from these checkpoints, ensuring data integrity and processing continuity without losing critical information. In scale down or up situations there could be a moment to reprocess data from last checkpoint. To reduce that amount we reduce the checkpointing interval to 10 seconds.
```

**Review by damccorm — APPROVED** [review]
Just had one last (tiny) suggestion

**Comment by talatuyarer** [comment]
@damccorm Looks like Master build is broken. fyi

**Comment by damccorm** [comment]
@talatuyarer which check? Looks like https://github.com/apache/beam/actions/workflows/beam_Publish_Website.yml is green at least

**Comment by talatuyarer** [comment]
But i dont see the post on blog page https://beam.apache.org/blog/

**Comment by damccorm** [comment]
The website publish job I linked to runs every 6 hours and hasn't run since we merged the PR

**Pull request merged**